WIP Geometry shaders

js6i opened the pull request:
Works in some cases.

cdavis5e committed:
Advertise the VK_EXT_shader_demote_to_helper_invocation extension.

As of macOS Big Sur and iOS/tvOS 14, the `discard_fragment()` function
in MSL is defined to have demote semantics; that is, fragment shader
output is discarded, but the fragment shader thread continues to run as
a helper invocation. This is very useful for Direct3D emulation, since
this is the semantic that HLSL `discard` has.

Signed-off-by: Chip Davis <chip@holochip.com>

maksym-pasichnyk commented on fetchDependencies:
```diff
@@ -304,7 +304,7 @@ if [ ! "$SPIRV_CROSS_ROOT" = "" ]; then
 	ln -sfn ${SPIRV_CROSS_ROOT} ${REPO_NAME}
 else
-	REPO_URL="https://github.com/KhronosGroup/${REPO_NAME}.git"
+	REPO_URL="https://github.com/js6i/${REPO_NAME}.git"
```
Maybe return KhronosGroup?

italomandara replied:
For the time being He needs to pull from his fork of SPIRV-CROSS a lot of GS work is done there too.

jakubnavratil commented:
Just asking if someone tryied this, does this solve Rocket League nameplates missing bug? Or possibly in future ...

italomandara commented:
> Just asking if someone tryied this, does this solve Rocket League nameplates missing bug? Or possibly in future ...

Yes among many other things, like unlocking the entire UE4 library of games and much more.

jakubnavratil commented:
> > Just asking if someone tryied this, does this solve Rocket League nameplates missing bug? Or possibly in future ...
> 
> Yes among many other things, like unlocking the entire UE4 game library and much more.

Thats great to hear. I can't imagine how much work this takes, I'm not really into graphical programming, but I hope this can be resolved this year. I will certainly try this on my Air when I have time. Thanks!

SuddenDeathMP commented:
> > Just asking if someone tryied this, does this solve Rocket League nameplates missing bug? Or possibly in future ...
> 
> Yes among many other things, like unlocking the entire UE4 game library and much more.

Did you actually tried to lunch any UE4 game with this branch?

italomandara commented:
> > > Just asking if someone tryied this, does this solve Rocket League nameplates missing bug? Or possibly in future ...
> > 
> > 
> > Yes among many other things, like unlocking the entire UE4 game library and much more.
> 
> Did you actually tried to lunch any UE4 game with this branch?

I did some tests, nothing works yet, this is a work in progress.

italomandara committed:
Allow to disable Explicit LOD Workaround

billhollings committed:
Merge pull request #1861 from italomandara/allow-disable-explicit-lod-workaround

Allow to disable Explicit LOD Workaround for Apple silicon.

billhollings committed:
Several simple maintenance fixes.

- Fix memory leak when waiting on timeline semaphores.
- For correctness, set VkPhysicalDeviceLimits::lineWidthGranularity to 1.
- Update MoltenVK to version 1.2.4.
- Update Whats_New.md document with recent changes.

billhollings committed:
Merge pull request #1865 from billhollings/timeline-semaphore-mem-leak

Several simple maintenance fixes.

billhollings committed:
Support BC compression on iOS/tvOS, where available.

- Add MVK_XCODE_14_3 macro to compile for iOS/tvOS 16.4 and above.
- Add support for BC compression on iOS/tvOS 16.4 and above where supported.
- Consolidate MVKPixelFormats::modifyMTLFormatCapabilities(mtlDev)
  and centralize querying MTLDevice format methods for all platforms.

billhollings committed:
Replace references to macOS 10.16 with 11.0, and standardize use
of whitespace in MVKPixelFormats::modifyMTLFormatCapabilities().

billhollings committed:
Merge pull request #1866 from billhollings/ios-support-bc-compression

Support BC compression on iOS/tvOS, where available.

billhollings committed:
Ensure shaders using PhysicalStorageBufferAddresses encode the associated MTLBuffer.

- MVKDevice track VkBuffers marked with VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT.
- Add SPIRVToMSLConversionResultInfo::usesPhysicalStorageBufferAddressesCapability
  to detect and track shaders that use PhysicalStorageBufferAddresses capability,
  and track such shader stages within pipeline.
- MVKResourcesCommandEncoderState encode usage of VkBuffers marked with
  VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT when pipeline uses
  PhysicalStorageBufferAddresses capability.
- Rename MVKResourcesCommandEncoderState::encodeArgumentBufferResourceUsage()
  to encodeResourceUsage().
- MVKDevice move some functions to public scope and remove friend classes.
- MVKDeviceMemory ensure _vkMemAllocFlags is always initialized (unrelated).
- Rename MVKFoundation template method contains() to mvkContains() (unrelated).

billhollings committed:
Merge pull request #1868 from billhollings/gpu-addr-buff-use

Ensure shaders using PhysicalStorageBufferAddresses encode the associated MTLBuffer.

billhollings committed:
Disable pipeline cache compression prior to macOS 10.15 and iOS/tvOS 13.0.

billhollings committed:
Support separate depth and stencil attachments during dynamic rendering.

- MVKRenderSubpass add separate getDepthFormat() & getStencilFormat(),
  and isDepthAttachmentUsed() & isStencilAttachmentUsed() and use
  instead of testing pixel format for depth and stencil components.
- Add MVKRenderingAttachmentIterator class to consistently iterate,
  and take actions, on the attachments in VkRenderingInfo to create
  synthetic MVKRenderPass and extract image views and clear colors.
- Remove mvkCreateRenderPass() and mvkCreateFramebuffer() in favor
  of additional constructors, and remove mvkGetDepthStencilFormat() in
  favor of retrieving formats for separate depth and stencil attachments.
- MVKRenderpass constructors reorganize order of adding attachments and
  subpasses, and connecting the two.
- Renmame MVKRenderPassAttachment to MVKAttachmentDescription.
- MVKPipeline reorganize member variables to minimize gaps in content
  and remove unnecessary _isRasterizingDepthStencil member var (unrelated).

IsaacMarovitz commented:
Any updates on this PR? :)

billhollings committed:
Fix Metal validation errors with dynamic rendering.

- MVKPipeline validate depth & stencil formats before setting frag shader depth
  and stencil builtins, and before setting formats in MTLRenderPipelineDescriptor.
- MVKCmdClearAttachments always set depth/stencil format if subpass includes
  a depth or stencil attachment, even if they are not being cleared.
- MVKRenderingAttachmentIterator add synthetic attachment if depth or stencil
  attachment is not provided, but image format supports both depth and stencil.

billhollings committed:
Merge pull request #1877 from billhollings/pipeline-compress-older-os-fix

Disable pipeline cache compression prior to macOS 10.15 and iOS/tvOS 13.0

billhollings committed:
Clear attachments support separate depth and stencil attachments.

- Also rename kMVKCachedColorAttachmentCount to kMVKMaxColorAttachmentCount and
  kMVKCachedViewportScissorCount to kMVKMaxViewportScissorCount and (unrelated)

billhollings committed:
Merge pull request #1878 from billhollings/dyn-rend-separate-depth-stencil

Support separate depth and stencil attachments during dynamic rendering

billhollings committed:
Deprecate the obsolete and non-standard VK_MVK_moltenvk extension.

The VK_MVK_moltenvk extension has never been brought inside Vulkan, and
the functions have never been supported by the Vulkan Loader and Layers.
Most of the functionality has long been replaced by the official
VK_metal_objects extension.

- Remove VK_MVK_moltenvk as an advertised extension.
- Refactor vk_mvk_moltenvk.h header file into separate headers files:
    - mvk_config.h - Valid public config functions
    - mvk_private_api.h - Valid development debugging functions used with care
    - mvk_deprecated_api.h - Formally deprecated functions.
- Retain skeleton vk_mvk_moltenvk.h header file for legacy compatibility only.
- Update documentation and header comments to explain changes.

billhollings committed:
Expose pointers to functions from removed VK_MVK_moltenvk extension.

- MVKInstance reorganize list of entry points for visual clarity (unrelated).
- Fixes to deprecation documentation (unrelated).
- mvkStringsAreEqual() add small performance optimization (unrelated).

billhollings committed:
Merge pull request #1881 from billhollings/deprecate-VK_MVK_moltenvk-extension

Deprecate the obsolete and non-standard VK_MVK_moltenvk extension.

billhollings committed:
Avoid Metal validation warning when depth component swizzled away.

- Set Metal depth store action to MTLStoreActionDontCare when
  depth attachment exists, but depth component has been swizzled away.

billhollings committed:
Merge pull request #1885 from billhollings/fix-mtl-valid-err-depth-swizzle

Avoid Metal validation warning when depth component swizzled away

billhollings committed:
Reinstate VK_MVK_moltenvk extension and add simplified config functions.

- Reinstate VK_MVK_moltenvk extension, but log warning message when it is enabled.
- Add vkGetMoltenVKConfiguration2MVK() and vkSetMoltenVKConfiguration2MVK()
  to set config without passing a dummy VkInstance, and deprecate
  vkGetMoltenVKConfigurationMVK() and vkSetMoltenVKConfigurationMVK().

billhollings committed:
Merge pull request #1886 from billhollings/reinstate-mvk-extn

Reinstate VK_MVK_moltenvk extension and add simplified config functions.

billhollings committed:
Remove vkGetMoltenVKConfiguration2MVK() and vkSetMoltenVKConfiguration2MVK().

Also undeprecate the original vkGet/SetMoltenVKConfigurationMVK().

In expectation of the upcoming VK_EXT_layer_settings extension, it is felt that
adding these additional functions at this time would be confusing to app devs.

billhollings committed:
Merge pull request #1889 from billhollings/rmv-config2

Remove vkGetMoltenVKConfiguration2MVK() and vkSetMoltenVKConfiguration2MVK()

billhollings committed:
Fix race condition when updating values in VkPastPresentationTimingGOOGLE.

- MVKPresentableSwapchainImage::presentCAMetalDrawable() and
  addPresentedHandler() pass MVKImagePresentInfo by value instead
  of reference, to avoid callbacks colliding with tracked
  MVKImagePresentInfos being cleared when
  MVKQueuePresentSurfaceSubmission is destroyed after it is run.

billhollings committed:
Merge pull request #1890 from billhollings/fix-display-timing-race-condition

Fix race condition when updating values in VkPastPresentationTimingGOOGLE.

billhollings committed:
GitHub CI improvements.

- Build one universal build, instead of per-platform.
- Upload this single build artifact to GitHub.
- Upgrade to v3 of action dependencies to remove Node.js deprecation warnings.
- Avoid use of deprecated set-output GitHub action command.
- Use macOS 13 and Xcode 14.3.
- README.md document access to binary artifacts.

billhollings committed:
Merge pull request #1893 from billhollings/ci-improvements

GitHub CI improvements.

Commit:
Add GitHub Actions workflow for automatically creating releases.

etang-cw committed:
Respect the bind point supplied to vkCmdBindDescriptorSets

billhollings committed:
Merge pull request #1895 from Steveice10/main

Add GitHub Actions workflow for automatically creating releases.

billhollings committed:
GitHub CI streamline uploaded artifact sizes.

- Reinstate per-platform upload artifacts.
- Don't upload shader converter binaries.

billhollings committed:
Merge pull request #1896 from billhollings/ci-reduce-artifact-sizes

GitHub CI streamline uploaded artifact sizes.

Commit:
Update release action artifact glob.

billhollings committed:
Merge pull request #1897 from Steveice10/main

Update release action artifact glob.

Commit:
Make the GitHub Actions release workflow more lenient.

billhollings committed:
Merge pull request #1900 from Steveice10/release-workflow-patch

Make the GitHub Actions release workflow more lenient.

billhollings committed:
Merge pull request #1898 from etang-cw/PushBindPoint

Respect the bind point supplied to vkCmdBindDescriptorSets / vkCmdPushDescriptorSets

etang-cw committed:
Fix reorder-ctor warning

ChibiDenDen committed:
Fix simulator feature discovery

Fix depth clamp and texture swizzle feature discovery on simulator builds

Both of these features are not supported by the iphone simulator
and leads to crashes when used

billhollings committed:
Merge pull request #1902 from etang-cw/FixReorderCtor

Fix reorder-ctor warning

billhollings committed:
Merge pull request #1903 from ChibiDenDen/simulator_feature_discovery_fix

Fix simulator feature discovery

billhollings committed:
Identify each unsupported device feature flag that the app attempts to be enable.

- Make MVKDevice::enableFeatures() functions into templates to pass struct type.
- Add mvkGetAddressOfFirstMember() to retrieve the address of the first member of
  a struct, taking into consideration whether the struct has a Vulkan pNext member.
- Add mvk::getTypeName() and mvk::getOrdinalSuffix() string functions.

billhollings committed:
Merge pull request #1905 from billhollings/id-unsupp-dvc-feat-flags

Identify each unsupported device feature flag that the app attempts to enable.

billhollings committed:
Ensure swapchain image presented time is always populated when requested.

If Metal reports zero presentedTime, and desired presentation
time has not been set by app, use the current time.

billhollings committed:
Merge pull request #1906 from billhollings/ensure-present-time

Ensure swapchain image presented time is always populated when requested.

billhollings committed:
Improve support for deviceUUID and deviceLUID.

- MTLDevice registryID is not constant across OS reboots,
  which is not conformant with deviceUUID requirements.
- Replace with combination of MTLDevice location, locationNumber,
  peerGroupID, and peerIndex, which should define uniqueness,
  and should be constant across OS reboots.
- Populate deviceLUID from MTLDevice registryID.

billhollings committed:
Do not fail on request for timestamp query pool that is too large.

- Report error, but do not fail on request for timestamp query pool
  that is too large for MTLCounterSampleBuffer.
- Change reported error to VK_ERROR_OUT_OF_DEVICE_MEMORY and clarify
  text of error reported when timestamp query pool is too large.
- Clarify error reported for occlusion query pool errors (unrelated).

billhollings committed:
Merge pull request #1910 from billhollings/fix-ts-qp

Do not fail on request for timestamp query pool that is too large.

billhollings committed:
Merge pull request #1908 from billhollings/fix-deviceUUID

Populate deviceUUID from MTLDevice location and peer group info.

etang-cw committed:
Check if shader compiled before adding it to a pipeline

Prevents Metal from aborting when you try to set a null function on the descriptor

billhollings committed:
Merge pull request #1911 from etang-cw/NoAbortOnCompilationFailure

Check if shader compiled before adding it to a pipeline

js6i commented:
Currently we're focused on preparing the next CrossOver release, there will be updates after it's out.

italomandara committed:
update SPIRV-Cross revision to include explicit LOD workaround regression fix

italomandara committed:
Revert "Allow to disable Explicit LOD Workaround"

This reverts commit 6f3a2b709fbfcfdf7dc077b80b12f02463cc07b1.

billhollings committed:
Merge pull request #1913 from italomandara/update-spirv-cross-revision

Update SPIRV-Cross revision to fix explicit LOD workaround regression

billhollings committed:
Update dependency libraries to match Vulkan SDK 1.3.250.

- Update Whats New document.

billhollings committed:
Merge pull request #1916 from billhollings/sdk-1.3.250

Update dependency libraries to match Vulkan SDK 1.3.250.

billhollings committed:
Accumulate render stages when a resource is used by multiple descriptor bindings.

billhollings committed:
Merge pull request #1918 from billhollings/accumulate-render-stages

Accumulate render stages when a resource is used by multiple descriptor bindings.

billhollings committed:
Advertise VK_KHR_depth_stencil_resolve extension on all devices.

Advertise VK_KHR_depth_stencil_resolve extension on early iOS devices,
since VK_RESOLVE_MODE_SAMPLE_ZERO_BIT is supported on all devices,
even if other resolve modes are not, and makes it consistent with
Vulkan 1.2 mandatory support for VK_RESOLVE_MODE_SAMPLE_ZERO_BIT.

billhollings committed:
Merge pull request #1920 from billhollings/always-advertise-VK_KHR_depth_stencil_resolve

Advertise VK_KHR_depth_stencil_resolve extension on all devices.

billhollings committed:
Ensure compute encoding state is marked dirty for non-dispatch commands.

The same compute encoder is used across dispatches and other commands,
which may override compute state, and end up breaking subsequent dispatches.

- Mark compute encoding state dirty when following commands,
  which use Metal compute encoders, are issued:
  - vkCmdCopyBuffer()
  - vkCmdClearColorImage()
  - vkCmdClearDepthStencilImage()
  - vkCmdFillBuffer()
  - vkCmdCopyQueryPoolResults()

- MVKCommandEncoder move marking compute state dirty from
  endCurrentMetalEncoding() to getMTLComputeEncoder().
- For efficiency, don't prematurely force end of query copy compute encoder
  used on renderpass end, in case compute dispatches follow.

- Update MoltenVK to 1.2.5 (unrelated).

billhollings committed:
Merge pull request #1922 from billhollings/fix-comp-enc-state

Ensure compute encoding state is marked dirty for non-dispatch commands.

billhollings committed:
Add support for VK_PRESENT_MODE_IMMEDIATE_KHR to macOS Cube demo.

- Only log performance stats on FPS logging if logging style is explicitly
  set to MVK_CONFIG_ACTIVITY_PERFORMANCE_LOGGING_STYLE_FRAME_COUNT (unrelated).

billhollings committed:
Merge pull request #1927 from billhollings/demo-support-immediate-present-mode

Add support for VK_PRESENT_MODE_IMMEDIATE_KHR to macOS Cube demo.

billhollings committed:
Support VK_PRESENT_MODE_IMMEDIATE_KHR if VkPresentTimeGOOGLE::desiredPresentTime is zero.

- [MTLDrawable presentAtTime:] syncs to display vsync. To support
  VK_PRESENT_MODE_IMMEDIATE_KHR while using VkPresentTimeGOOGLE::presentID,
  only call presentAtTime: if VkPresentTimeGOOGLE::desiredPresentTime has
  been explicitly set to a non-zero value.
- Clarify initially clearing MVKImagePresentInfo to all zeros.

rcombs committed:
update SPIRV-Cross revision to include function constant fix; closes #1941

billhollings committed:
Merge pull request #1942 from rcombs/spirv-cross-fix

Update SPIRV-Cross revision to include function constant fix

cdavis5e committed:
Support the `VK_EXT_4444_formats` extension.

This turned out to be a little bit more involved than I had hoped. But,
with this, we can now use the `VK_FORMAT_A4R4G4B4_UNORM_PACK16` and
`VK_FORMAT_A4B4G4R4_UNORM_PACK16` formats from shaders, use them as blit
sources, and even clear them. Storage images and render targets of these
formats aren't supported, however. To support the latter would require
the insertion of a swizzle into the fragment shader before returning.
The former cannot be reasonably supported.

billhollings committed:
Merge pull request #1935 from cdavis5e/shader-demote-to-helper

Advertise the VK_EXT_shader_demote_to_helper_invocation extension.

billhollings committed:
Merge pull request #1936 from billhollings/present-imm-mode-unless-present-time

Support VK_PRESENT_MODE_IMMEDIATE_KHR if VkPresentTimeGOOGLE::desiredPresentTime is zero.

RaafatAkkad committed:
Allows maximizing the concurrent executing compilation tasks.

https://developer.apple.com/videos/play/wwdc2023/10127/?time=540
On an M1 Pro Macbook Pro 16" maximumConcurrentCompilationTaskCount goes from 2 to 8

billhollings committed:
Merge pull request #1944 from RaafatAkkad/main

Allows maximizing the concurrent executing compilation tasks.

billhollings committed:
Allow both renderPass and VkPipelineRenderingCreateInfo to be missing.

This avoids crashes in CTS when creating a graphics pipeline.

billhollings committed:
Merge pull request #1946 from billhollings/no-rendpass-no-dyn-rend-info

Allow both renderPass and VkPipelineRenderingCreateInfo to be missing.

billhollings committed:
Don't call MTLCreateSystemDefaultDevice() on every swapchain creation.

- When forcing the window system to use the same high-power GPU as the app,
  move the call to MTLCreateSystemDefaultDevice() to MVKDevice constructor,
  instead of MVKDevice::createSwapchain(), and test whether the
  VK_KHR_swapchain extension is enabled to determine the need to swap GPUs.

- After calling MTLCreateSystemDefaultDevice() the GPU will already be
  the same high-power GPU, so remove attempting to replace the MTLDevice.

- Remove MVKPhysicalDevice::replaceMTLDevice() as no longer used.

- Remove many unnecessary inline declarations in MVKDevice.h (unrelated).

billhollings committed:
Log more info about SPIR-V to MSL conversion errors.

Remove unhelpful "Shader module does not contain
an entry point named 'main0'" error message.

billhollings committed:
Merge pull request #1951 from billhollings/log-shdr-conv-err-info

Log more info about SPIR-V to MSL conversion errors.

billhollings committed:
Merge pull request #1949 from billhollings/swap-system-GPU-fix

Don't call MTLCreateSystemDefaultDevice() on every swapchain creation.

AntarticCoder committed:
Implemented Deferred Host Operations

Implemented deferred host operations in this commit. It was pretty simple with nothing Metal specific. I am a bit concerned on the return types of MVKDeferredOperation::join and about how to store the operation to be executed for later.

AntarticCoder committed:
Fixed comments for deferred operations

Co-authored-by: Chip Davis <cdavis5x@gmail.com>

AntarticCoder committed:
Made some fixes for Deferred Operations

The changes are as follows:

* Moved the code around to fit with the ordering system
* Added a function to get available cpu cores
* Renamed variables with _ in front of them
* Added mutexes and lock guards for the getters and setters of the max concurrency and result variables
* Made max concurrency dynamic by returning 0 when the operation is finished

AntarticCoder committed:
Push to Compile MVKSmallVector with void*

Pushing to Github to compile with Github Actions to see if MVKVectors with void* can work

AntarticCoder committed:
C-Style Arrays instead of MVKSmallVector

This commit changes the MVKSmallVector to a C-Style array because the parameters array will not have to resize, and it seems that a C-Style array is just as capable of doing the job.

billhollings committed:
Merge pull request #1954 from AntarticCoder/khr-deferred-operation

Implemented Deferred Host Operations

cdavis5e committed:
Support the `VK_EXT_calibrated_timestamps` extension.

This extension has a direct Metal equivalent in the
`-[MTLDevice sampleTimestamps:gpuTimestamp:]` method. However, that
method returns CPU timestamps in the Mach absolute time domain, which is
*not* that of `CLOCK_MONOTONIC_RAW` but of `CLOCK_UPTIME_RAW`. The
function that corresponds to `CLOCK_MONOTONIC_RAW` is
`mach_continuous_time()`. Therefore, this implementation uses the
`mach_continuous_time()` function for the CPU timestamp. Perhaps we
should lobby the WG for `VK_TIME_DOMAIN_CLOCK_UPTIME_RAW_EXT`.

ifiddynine committed:
Support building for visionOS

Note: Internal naming is xrOS, but publicly it should be referred to as visionOS

ifiddynine committed:
Merge branch 'main' of https://github.com/KhronosGroup/MoltenVK into visionOS

ifiddynine committed:
Fix MVK_MACOS_OR_IOS_OR_VISIONOS check

ifiddynine committed:
Fix build with latest main

cdavis5e committed:
Merge pull request #1955 from cdavis5e/4444-formats

Support the `VK_EXT_4444_formats` extension.

ifiddynine committed:
Fix mvkOSVersionIsAtLeast on Vision

cdavis5e committed:
Support `VK_EXT_shader_subgroup_ballot` and `VK_EXT_shader_subgroup_vote`.

Update SPIRV-Cross to pull in the needed support for these extensions.

cdavis5e committed:
Advertise the `VK_KHR_shader_non_semantic_info` extension.

This just provides support for the `SPV_KHR_non_semantic_info`
extension, which supports extended instruction sets that do not affect
the semantics of a SPIR-V shader (e.g. debug info). SPIRV-Cross already
handles these instruction sets, so no additional work is required on our
part to support this extension.

billhollings committed:
Merge pull request #1958 from cdavis5e/calibrated-timestamps

Support the `VK_EXT_calibrated_timestamps` extension.

billhollings committed:
Merge pull request #1960 from cdavis5e/shader-non-semantic-info

Advertise the `VK_KHR_shader_non_semantic_info` extension.

billhollings committed:
Merge pull request #1961 from cdavis5e/legacy-subgroup

Support `VK_EXT_shader_subgroup_ballot` and `VK_EXT_shader_subgroup_vote`.

billhollings committed:
Add support for VK_PRIMITIVE_TOPOLOGY_TRIANGLE_FAN.

To reduce complexity and repetitive copy-pasted spaghetti code,
the design approach here was to implement triangle fan conversion on
MVKCmdDrawIndexedIndirect, as the most general of the draw commands,
and then populate and invoke a synthetic MVKCmdDrawIndexedIndirect
command from the other draw commands.

- Rename pipeline factory shader cmdDrawIndexedIndirectMultiviewConvertBuffers()
  to cmdDrawIndexedIndirectConvertBuffers, and in addition to original support
  for modifying indirect content to support multiview, add support for
  converting triangle fan indirect content and indexes to triangle list.
- Modify MVKCmdDrawIndexedIndirect to track need to convert triangle fans
  to triangle list, and invoke kernel function when needed.
- Modify MVKCmdDraw, MVKCmdDrawIndexed, and MVKCmdDrawIndirect to populate
  and invoke a synthetic MVKCmdDrawIndexedIndirect command to convert triangle
  fans to triangle lists.
- Add pipeline factory shader cmdDrawIndirectPopulateIndexes() to convert
  non-indexed indirect content to indexed indirect content.
- MVKCmdDrawIndexedIndirect add support for zero divisor vertex buffers
  potentially coming from MVKCmdDraw and MVKCmdDrawIndexed.

- Rename pipeline factory shader cmdDrawIndexedIndirectConvertBuffers()
  to cmdDrawIndexedIndirectTessConvertBuffers() so it will be invoked from
  MVKCommandEncodingPool::getCmdDrawIndirectTessConvertBuffersMTLComputePipelineState()
  (unrelated).

billhollings committed:
Fix unreachable code in MVKDeferredOperation::join().

- Fix unreachable code in MVKDeferredOperation::join().
- Refactor code so deferred functions call back to MVKDeferredOperation
  instance to update current status, and deferred function execution
  returns result of individual thread execution.
- MVKDeferredOperation use MVKSmallVector for _functionParameters.
- MVKDeferredOperation use a single mutex lock.
- Add additional comments explaining design to developers of
  future extensions that use deferred operations.

billhollings committed:
Merge pull request #1964 from billhollings/fix-unreachable-code

Fix unreachable code in MVKDeferredOperation::join().

billhollings committed:
Merge pull request #1962 from billhollings/tri-fans

Add support for VK_PRIMITIVE_TOPOLOGY_TRIANGLE_FAN.

AntarticCoder committed:
Added MSL Version 3.1 for switch case

In this commit, I've added support for Xcode 15, and added a case for MSL version 3.1. I added this because I noticed xcode was throwing some warnings about an unhandled switch case.

AntarticCoder committed:
Completed the support for MSL 3.1 enum

This commit adds some extra code that was needed to define to MSL 3.1 enum that was not included in the first commit. Based on PR: #1940

billhollings committed:
Merge pull request #1966 from AntarticCoder/msl-3-1

Added Enum for MSL Version 3.1

cdavis5e committed:
Support the `VK_EXT_pipeline_creation_feedback` extension.

This provides feedback that indicates:
* how long it took to compile each shader stage and the pipeline as a
  whole;
* whether or not the pipeline or any shader stage were found in any
  supplied pipeline cache; and
* whether or not any supplied base pipeline were used to accelerate
  pipeline creation.

This is similar to the performance statistics that MoltenVK already
collects.

Since we don't use any supplied base pipeline at all, this
implementation never sets
`VK_PIPELINE_CREATION_FEEDBACK_BASE_PIPELINE_ACCELERATION_BIT`. However,
I've identified several places where we could probably use the base
pipeline to accelerate pipeline creation. One day, I should probably
implement that.

Likewise, because we don't yet support using `MTLBinaryArchive`s,
`VK_PIPELINE_CREATION_FEEDBACK_APPLICATION_PIPELINE_CACHE_HIT_BIT` is
never set on the whole pipeline, though it *is* set for individual
stages, on the assumption that any shader found in a cache is likely to
be found in Metal's own implicit cache.

In this implementation, shader stage compilation time includes any time
needed to build the `MTLComputePipelineState`s needed for vertex and
tessellation control shaders in tessellated pipelines.

This patch also changes compilation of the vertex stage
`MTLComputePipelineState`s in tessellated pipelines to be eager instead
of lazy. We really ought to have been doing this anyway, in order to
report pipeline failures at creation time instead of draw time. I'm not
happy, though, that we now pay the cost of all three pipeline states all
the time, instead of just the ones that are used.

This also gets rid of some fields of `MVKGraphicsPipeline` that were
only used during pipeline construction, which should save some memory,
particularly for apps that create lots of pipelines.

cdavis5e committed:
MVKDevice: Don't enable sample LoD depth array workaround for macOS Sonoma and up.

Apple have indicated to me that they have fixed the bug. I've confirmed
this.

cdavis5e committed:
Support the `VK_KHR_incremental_present` extension.

This extension allows apps to provide a hint to the presentation engine
indicating which parts of the surface need updating. To provide this
hint, we call `-[CALayer setNeedsDisplayInRect:]`, which indicates that
only the given rectangle needs updating.

I'm not sure if this will have any effect, especially if
`CAMetalLayer.presentsWithTransaction` is `NO`. Luckily for us, this is
only a hint, and it is permissible for the presentation engine to do
nothing with the hint.

The tests don't work because they apparently can't handle
`VK_SUBOPTIMAL_KHR` being returned.

billhollings committed:
Merge pull request #1969 from cdavis5e/sonoma-depth-array-lod

MVKDevice: Don't enable sample LoD depth array workaround for macOS S…

billhollings committed:
Merge pull request #1968 from cdavis5e/pipeline-creation-feedback

Support the `VK_EXT_pipeline_creation_feedback` extension.

billhollings committed:
Merge pull request #1972 from cdavis5e/incremental-present

Support the `VK_KHR_incremental_present` extension.

billhollings committed:
Merge branch 'visionOS' of https://github.com/nginetechnologies/MoltenVK into nginetechnologies-visionOS

billhollings committed:
Fix visionOS build errors and warnings where possible.

- Remove visionOS from multi-platform builds because it
  requires Xcode 15+ and will abort a multi-platform build.
- Define TARGET_OS_XR for older SDK's.
- A number of SDK deprecation warnings remain when building for visionOS.
  These cannot be removed without significant refactoring.
- Build visionOS dependencies for Release build by default.
- Fix local variable initialization warning (unrelated).

billhollings committed:
Merge pull request #1976 from billhollings/nginetechnologies-visionOS

Support building for visionOS platform

billhollings committed:
Drop official support for using Xcode 11 to build MoltenVK.

- Remove Xcode 11 build from GitHub CI.
- Leave MVK_XCODE_12 guards in place to allow devs to possibly continue to
  attempt to build existing MoltenVK code using Xcode 11, even though it's
  not officially supported. Such devs may have to add their own additional
  MVK_XCODE_12 guards for any Xcode 12 API features added after this change.

billhollings committed:
Merge pull request #1977 from billhollings/drop-support-xcode-11

Drop official support for using Xcode 11 to build MoltenVK.

oscarbg commented:
Seems this work, is part of the present crossover 23 beta1 release which says:
“We’re pleased to announce that this release also includes initial support for geometry shaders and transform feedback on macOS”

mbechard committed:
Fix regression caused by #1922

Needed to complete fix for #1874

We can't wait until getMTLComputeEncoder() is called to dirty the state,
because this call will be avoided by dirty checks themselves.
Those checks are comparing against leftover and now incorrect state since
the previous encoder has already ended.
It needs to be dirtied on encoder end.

billhollings committed:
Merge pull request #1980 from mbechard/main

Fix regression caused by #1922

billhollings committed:
Fix bad access regression to prematurely-deallocated MTLFunction.

When compiling tessellation vertex shaders, MVKGraphicsPipeline
pass array of MVKMTLFunction instead of MTLFunctions to retain
MTLFunctions for duration of processing.

billhollings committed:
Merge pull request #1981 from billhollings/fix-bad-access-regression

Fix bad access regression to prematurely-deallocated MTLFunction.

stefand committed:
Don't fetch a submodule if the commit is already known

This allows rebuilding without an internet connection and avoids random
build failures on flaky connections.

billhollings committed:
Merge pull request #1982 from stefand/main

Don't fetch a submodule if the commit is already known

billhollings committed:
Fix sync delay between calls to vkQueueSubmit() on non-Apple-Silicon devices.

The [MTLDevice sampleTimestamps:gpuTimestamp:] function turns out to be
synchronized with other queue activities, and can block GPU execution
if it is called between MTLCommandBuffer submissions. On non-Apple-Silicon
devices, it was called before and after every vkQueueSubmit() submission,
to track the correlation between GPU and CPU timestamps, and was delaying
the start of GPU work on the next submission (on Apple Silicon, both
CPU & GPU timestamps are specified in nanoseconds, and the call was bypassed).

Move timestamp correlation from vkQueueSubmit() to
vkGetPhysicalDeviceProperties(), where it is used to update
VkPhysicalDeviceLimits::timestampPeriod on non-Apple-Silicon devices.

Delegate MVKPhysicalDevice::getProperties(VkPhysicalDeviceProperties2*)
to MVKPhysicalDevice::getProperties(VkPhysicalDeviceProperties*), plus
minimize wasted effort if pNext is empty (unrelated).

Move the declaration of several MVKPhysicalDevice member structs to
potentially reduce member spacing (unrelated).

billhollings committed:
Ensure Xcode simulator always uses 256B buffer alignment.

Xcode simulator always requires 256B buffer alignment, even when running
on Apple Silicon. Previously, it was assumed that Apple Silicon would use
it's native 16B buffer alignment.

billhollings committed:
Merge pull request #1992 from billhollings/fix-submit-sync-delay

Fix sync delay between calls to vkQueueSubmit() on non-Apple-Silicon devices.

billhollings committed:
Merge pull request #1993 from billhollings/fix-sim-buff-align

Ensure Xcode simulator always uses 256B buffer alignment.

billhollings committed:
Update dependency libraries to match Vulkan SDK 1.3.261.

- In MoltenVK Xcode projects, set iOS & tvOS deployment targets to 12.0,
  to avoid warnings while building MoltenVK.
- Add DYLD_LIBRARY_PATH to runcts script, to ensure Vulkan and MoltenVK
  libraries are found during CTS runs.
- Update Whats_New.md and MoltenVK_Runtime_UserGuide.md documents.

billhollings committed:
Merge pull request #1994 from billhollings/sdk-1.3.261

Update dependency libraries to match Vulkan SDK 1.3.261.

billhollings committed:
Fix VkPhysicalDeviceLimits::timestampPeriod calculations on Intel GPU.

- Guard against Intel returning zero values for CPU & GPU timestamps.
- Apply lowpass filter on timestampPeriod updates, to avoid wild temporary
  changes, particularly at startup before GPU has been really exercised.

billhollings committed:
Merge pull request #1997 from billhollings/fix-Intel-timestampPeriod

Fix VkPhysicalDeviceLimits::timestampPeriod calculations on Intel GPU.

billhollings committed:
Fix rare case where vertex attribute buffers are not bound to Metal.

In the rare case where vertex attribute buffers are bound to MVKCommandEncoder,
are not used by first pipeline, but are used by a subsequent pipeline, and no
other bindings are changed, the MVKResourcesCommandEncoderState will not appear
to be dirty to the second pipeline, and the buffer will not be bound to Metal.

When reverting a binding to dirty if it is not used by a pipeline, also revert
the enclosing MVKResourcesCommandEncoderState to dirty state.

Update MoltenVK to version 1.2.6 (unrelated).

billhollings committed:
Merge pull request #2001 from billhollings/fix-unbound-vtx-attr-buff

Fix rare case where vertex attribute buffers are not bound to Metal.

js6i committed:
Implement geometry shaders using mesh pipelines.

js6i committed:
Temporarily switch SPIRV-Cross repository.

alphavector commented:
I built libMoltenVK.dylib with @js6i's latest changes and I get the error `DX11 feature level 10.0 is required to run the engine.`

```sh
> xcodebuild -version
Xcode 15.0
Build version 15A240d
> sw_vers
ProductName: macOS
ProductVersion: 14.0
BuildVersion: 23A339
```
I also replaced MACOSX_DEPLOYMENT_TARGET
```diff
diff --git a/ExternalDependencies.xcodeproj/project.pbxproj b/ExternalDependencies.xcodeproj/project.pbxproj
index f92a13b8..6496f6ed 100644
--- a/ExternalDependencies.xcodeproj/project.pbxproj
+++ b/ExternalDependencies.xcodeproj/project.pbxproj
@@ -5407,7 +5407,7 @@
 				GCC_WARN_UNUSED_VARIABLE = NO;
 				IPHONEOS_DEPLOYMENT_TARGET = 9.0;
 				MACH_O_TYPE = staticlib;
-				MACOSX_DEPLOYMENT_TARGET = 10.11;
+				MACOSX_DEPLOYMENT_TARGET = 10.14;
 				SKIP_INSTALL = YES;
 				TVOS_DEPLOYMENT_TARGET = 9.0;
 			};
@@ -5456,7 +5456,7 @@
 				GCC_WARN_UNUSED_VARIABLE = NO;
 				IPHONEOS_DEPLOYMENT_TARGET = 9.0;
 				MACH_O_TYPE = staticlib;
-				MACOSX_DEPLOYMENT_TARGET = 10.11;
+				MACOSX_DEPLOYMENT_TARGET = 10.14;
 				SKIP_INSTALL = YES;
 				TVOS_DEPLOYMENT_TARGET = 9.0;
 				VALIDATE_PRODUCT = YES;
diff --git a/MoltenVK/MoltenVK.xcodeproj/project.pbxproj b/MoltenVK/MoltenVK.xcodeproj/project.pbxproj
index 650a8219..b50158d6 100644
--- a/MoltenVK/MoltenVK.xcodeproj/project.pbxproj
+++ b/MoltenVK/MoltenVK.xcodeproj/project.pbxproj
@@ -1594,7 +1594,7 @@
 				);
 				IPHONEOS_DEPLOYMENT_TARGET = 9.0;
 				MACH_O_TYPE = staticlib;
-				MACOSX_DEPLOYMENT_TARGET = 10.11;
+				MACOSX_DEPLOYMENT_TARGET = 10.14;
 				MTL_ENABLE_DEBUG_INFO = YES;
 				MVK_SKIP_DYLIB = "";
 				"MVK_SKIP_DYLIB[sdk=appletvsimulator*]" = YES;
@@ -1664,7 +1664,7 @@
 				);
 				IPHONEOS_DEPLOYMENT_TARGET = 9.0;
 				MACH_O_TYPE = staticlib;
-				MACOSX_DEPLOYMENT_TARGET = 10.11;
+				MACOSX_DEPLOYMENT_TARGET = 10.14;
 				MTL_ENABLE_DEBUG_INFO = NO;
 				MVK_SKIP_DYLIB = "";
 				"MVK_SKIP_DYLIB[sdk=appletvsimulator*]" = YES;
diff --git a/MoltenVKShaderConverter/MoltenVKShaderConverter.xcodeproj/project.pbxproj b/MoltenVKShaderConverter/MoltenVKShaderConverter.xcodeproj/project.pbxproj
index 924a8987..9ddbf11b 100644
--- a/MoltenVKShaderConverter/MoltenVKShaderConverter.xcodeproj/project.pbxproj
+++ b/MoltenVKShaderConverter/MoltenVKShaderConverter.xcodeproj/project.pbxproj
@@ -632,7 +632,7 @@
 				);
 				IPHONEOS_DEPLOYMENT_TARGET = 9.0;
 				MACH_O_TYPE = staticlib;
-				MACOSX_DEPLOYMENT_TARGET = 10.11;
+				MACOSX_DEPLOYMENT_TARGET = 10.14;
 				PRODUCT_NAME = MoltenVKShaderConverter;
 				SKIP_INSTALL = YES;
 				TVOS_DEPLOYMENT_TARGET = 9.0;
@@ -691,7 +691,7 @@
 				);
 				IPHONEOS_DEPLOYMENT_TARGET = 9.0;
 				MACH_O_TYPE = staticlib;
-				MACOSX_DEPLOYMENT_TARGET = 10.11;
+				MACOSX_DEPLOYMENT_TARGET = 10.14;
 				PRODUCT_NAME = MoltenVKShaderConverter;
 				SKIP_INSTALL = YES;
 				TVOS_DEPLOYMENT_TARGET = 9.0;
```



Build of the latest release version works

js6i commented:
> I built libMoltenVK.dylib with @js6i's latest changes and I get the error `DX11 feature level 10.0 is required to run the engine.`

It's probably missing shaderCullDistance and, if you're running wined3d, pipelineStatisticsQuery features, didn't you apply a hack adding those on top of release to make it work?

alphavector commented:
I'm using dxvk. I'm unlikely to apply any hacks, I'm a regular user who just follows this MR :D. I looked at the game launch logs with version 1.2.3, which I built myself with hacks to properly display shaders instead of black textures and my game runs fine

And used your changes + merged the main branch

I tried to compare the logs, but it's not obvious to me what the difference is and what went wrong

alphavector commented:
I can help with the debug if needed
Or it's worth waiting until this mr is fully ready and it's too early to test anything yet

stossy11 commented:
Any News on this pr as if I try to build it I get this error when trying to fetch the dependency "SPIRV-Cross"

fatal: reference is not a tree: 7a1e3e7586b9b10a6286cad912940b7a501e93ea

italomandara commented:
> Any News on this pr as if I try to build it I get this error when trying to fetch the dependency "SPIRV-Cross"
> 
> fatal: reference is not a tree: 7a1e3e7586b9b10a6286cad912940b7a501e93ea

That's because you don't have access to this repo `REPO_URL="https://github.com/js6i/${REPO_NAME}.git"`
you could do a `git remote add`

okaku123 commented:
> > Any News on this pr as if I try to build it I get this error when trying to fetch the dependency "SPIRV-Cross"
> > fatal: reference is not a tree: 7a1e3e7586b9b10a6286cad912940b7a501e93ea
> 
> That's because you don't have access to this repo `REPO_URL="https://github.com/js6i/${REPO_NAME}.git"` you could do a `git remote add`

no seen any branch with this id, and there is no latest code related to geometry-shaders branch(https://github.com/js6i/SPIRV-Cross/tree/geometry-shaders).

<img width="875" alt="image" src="https://github.com/user-attachments/assets/6570494f-844a-4712-b60a-349787c9ec21">
<img width="559" alt="image" src="https://github.com/user-attachments/assets/64e750b4-d189-4ba7-a859-b991b5b788e2">

The latest code seems to be missing here.How can I get this branch? I want to try its code.

ovvldc commented:
So Alyssa Rosenzweig seems to have added this feature to her Linux driver in the mean time. Is that enough guidance for someone to make this work? See https://rosenzweig.io/blog/aaa-gaming-on-m1.html

C4ndyF1sh commented:
any updates on this branch?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/MoltenVK?pullRequest=1815) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/MoltenVK?pullRequest=1815) before we can accept your contribution.<br/>**1** out of **5** committers have signed the CLA.<br/><br/>:white_check_mark: js6i<br/>:x: cdavis5e<br/>:x: AntarticCoder<br/>:x: mbechard<br/>:x: stefand<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/MoltenVK?pullRequest=1815) it.</sub>

IsaacMarovitz commented:
@js6i Any chance of a rebase at least?
